Use static type information to restore type tags

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26271 [JIT] add type refinements for isinstance checks
* #26270 [jit] refactor extra sugared values
* #26269 [jit] add dynamic isinstance
* #26268 Allow types as node attributes
* #26145 [JIT] add CondValue to unify refinements and code emission
* #26061 [jit] Refactor emitIsInstance
* **#25447 Use static type information to restore type tags**
* #25439 Add ivalue::type(), part 1

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

Differential Revision: [D17127872](https://our.internmc.facebook.com/intern/diff/D17127872)

zdevito committed:
Use static type information to restore type tags

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

zdevito committed:
Update on "Use static type information to restore type tags"

Use static type information to restore type tags

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

gh-metadata: pytorch pytorch 25447 gh/zdevito/102/head

zdevito committed:
Update on "Use static type information to restore type tags"

Use static type information to restore type tags

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

emitModuleHook now has a check that the tags in saved/load modules
are the same.

gh-metadata: pytorch pytorch 25447 gh/zdevito/102/head

zdevito committed:
Update on "Use static type information to restore type tags"

Use static type information to restore type tags

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

emitModuleHook now has a check that the tags in saved/load modules
are the same.

gh-metadata: pytorch pytorch 25447 gh/zdevito/102/head

zdevito committed:
Update on "Use static type information to restore type tags"

Use static type information to restore type tags

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

emitModuleHook now has a check that the tags in saved/load modules
are the same.

gh-metadata: pytorch pytorch 25447 gh/zdevito/102/head

zdevito committed:
Update on "Use static type information to restore type tags"

Use static type information to restore type tags

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

emitModuleHook now has a check that the tags in saved/load modules
are the same.

gh-metadata: pytorch pytorch 25447 gh/zdevito/102/head

zdevito committed:
Update on "Use static type information to restore type tags"

Use static type information to restore type tags

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

emitModuleHook now has a check that the tags in saved/load modules
are the same.

gh-metadata: pytorch pytorch 25447 gh/zdevito/102/head

suo commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -551,12 +551,125 @@ void Pickler::pushTuple(const IValue& ivalue) {
   }
 }
 
+static void restoreAccurateTypeTagsIfPossible(const IValue& root) {
```
Can we add a comment explaining what this is for? What happens if restoring accurate type tags is *not* possible? The name alone is so tantalizing and full of mystery!

suo reviewed (approved):
ngnt, just want more comments for posterity

zdevito committed:
Update on "Use static type information to restore type tags"

Use static type information to restore type tags

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

emitModuleHook now has a check that the tags in saved/load modules
are the same.

gh-metadata: pytorch pytorch 25447 gh/zdevito/102/head

zdevito committed:
Update on "Use static type information to restore type tags"

Use static type information to restore type tags

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

emitModuleHook now has a check that the tags in saved/load modules
are the same.

gh-metadata: pytorch pytorch 25447 gh/zdevito/102/head

zdevito committed:
Update on "Use static type information to restore type tags"

Use static type information to restore type tags

When we unpickle IValues, we lose type information for List[T]
and Dict[K, V]. We can restore this information using the static
type information contained in the top-level Module/Class type.

This ensures that even after serialization we can always get the
dynamic type of an ivalue using its type() method.

emitModuleHook now has a check that the tags in saved/load modules
are the same.

gh-metadata: pytorch pytorch 25447 gh/zdevito/102/head
